### DISCUSSION TITLE
Export WASM to folder for self-publishing on personal web site

### DISCUSSION THREAD

**Discussion by astrowonk** [marimo-team/marimo#1078] [Ideas]
marimo.app is very cool and I was excited to see the WASM support. But, WASM apps should be able to just be turned into a folder than can deployed on any web server (nginx, apache, etc.) [Shiny Live](https://github.com/posit-dev/py-shinylive) for example has an easy `shinylive export myapp site` command that turns the app into a folder I can upload to my nginx server.

Whereas for now, wasm notebooks seem to require using `marimo.app`. Is there a plan to have an export option someday? Thanks!

**Comment by akshayka** [comment]
This is a great idea, and has been requested a few times. We very well may do this in the future.

In the meantime, can you get by with iframing marimo.app (https://docs.marimo.io/guides/wasm.html#embedding)?

**Reply by astrowonk** [reply]
Hmm, I don't think so. I keep most of my complex code in separate python files, that I import into the Marimo notebook to use/visualize. These functions/classes are not defined in the notebook itself. Similarly at some point I'd want to [read data from a local file](https://shinylive.io/py/examples/#read-local-csv), assuming that works.

**Reply by dmadisetti** [reply]
@astrowonk Support for external files would require some work to the wasm virtual file system. But a "caching" system is being considered, that could run code on export, but then save the computed value to just be loaded in the export. This would also be faster to the end user

**Comment by jettil** [comment]
That would be a nice feature! I Think it would solve many data security concerns and help publish marimo notebooks to non tech colleagues

**Comment by aszenz** [comment]
Lot of private data sources cannot be loaded from  cross origin, so embedding an iframe isn't useful.

Starboard notebook allows easy embedding into any webpage using a plain js+css file: https://github.com/gzuidhof/starboard-notebook

**Comment by akshayka** [comment]
@aszenz , @jettil , and @astrowonk : we now have the pieces to make this possible, with [marimo islands](https://docs.marimo.io/guides/wasm.html#islands). In fact this is possible today using islands yourself, but I recognize that's more cognitive overhead than a CLI command. We'll add support for this.

**Reply by akshayka** [reply]
Issue: https://github.com/marimo-team/marimo/issues/1882

**Reply by aszenz** [reply]
> @aszenz , @jettil , and @astrowonk : we now have the pieces to make this possible, with [marimo islands](https://docs.marimo.io/guides/wasm.html#islands). In fact this is possible today using islands yourself, but I recognize that's more cognitive overhead than a CLI command. We'll add support for this.

Is it possible to have the notebook editor in the islands, I thought it's only meant for rendering the content

**Reply by dmadisetti** [reply]
@aszenz Islands allows you to have editable cells- but yes it is more of an export than a notebook experience. The cells are still reactive and powered by wasm- but you can't create or delete cells as is.

You could hack up a solution particular to your islands integration if you wanted to make it more "notebook" like, but I think it'd just add additional surface area for core marimo - especially since the notebook experience is so well integrated.

**Reply by colluca** [reply]
> @aszenz , @jettil , and @astrowonk : we now have the pieces to make this possible, with [marimo islands](https://docs.marimo.io/guides/wasm.html#islands). In fact this is possible today using islands yourself, but I recognize that's more cognitive overhead than a CLI command. We'll add support for this.

This still doesn't allow "storing code in separate python files" as @astrowonk mentioned, is that correct? I found myself in a similar need when experimenting with WASM notebooks.

**Comment by mscolnick** [comment]
Good news! This is now possible in marimo [0.10.0](https://github.com/marimo-team/marimo/releases/tag/0.10.0), here are the [docs](https://docs.marimo.io/guides/exporting.html#export-to-wasm-powered-html). (related [PR](https://github.com/marimo-team/marimo/pull/3134)).

```bash
# export as readonly, with code locked
marimo export html-wasm notebook.py -o output_dir --mode run
# export as an editable notebook
marimo export html-wasm notebook.py -o output_dir --mode edit
```

The exported HTML file will run your notebook using WebAssembly, making it completely self-contained and executable in the browser. This means users can interact with your notebook without needing Python or marimo installed.

Options:

- `--mode`: Choose between `run` (read-only) or `edit` (allows editing)
- `--output`: Directory to save the HTML and required assets

**Comment by ghost** [comment]
Going off the original question of needing to publish on a personal website using Apache,
How does one include all their dependencies when using the above? The assets folder doesn't include the packages used in my notebook, so how does this work?  
(only asking because the filesize of assets seems smaller than I would expect considering the packages needed)

**Reply by mscolnick** [reply]
correct, we do not download all the packages. but they are loaded on startup from pypi

**Reply by ghost** [reply]
A continuation from https://github.com/marimo-team/marimo/issues/3859  , the pipeline I'm hoping for is to include/download all required packages locally for the end user when the page loads for the first time using the internet, and then have the ability to save the web page to the end users device so that subsequent visits to the webpage do not need the internet since its already loaded. But im unclear if this is possible to accomplish if I dont want the source code visible to the end-user. So not sure if things like apache, nginx are the correct methods.
